crates/energy/benches/nearby.rs

```diff
@@ -0,0 +1,152 @@
+
+impl Clock {
+    fn inc(&mut self) {
+        self.0 += 0.008 // 125 updates a "second"
+    }
+}
+
+fn update_index(
+    mut index: ResMut<EntityIndex>,
+    moved: Query<(Entity, &Transform), Changed<Transform>>,
+) {
+    for (entity, transform) in moved.iter() {
+        let position = Isometry::new(
+            transform.translation.into(),
+            transform.rotation.to_scaled_axis().into(),
+        );
+        index.update(entity, position);
+    }
+}
+
+fn init_world_with_entities_moving(world: &mut World, num_entities: u32) {
+    let mut index = EntityIndex::new();
+    let points = load_points(
+        num_entities.try_into().unwrap(),
+        MAP_SIZE - DISTANCE_FROM_MAP_EDGE,
+    );
+
+    for (i, point) in points.into_iter().enumerate() {
+        let point_msl = point.to_msl();
+
+        let collider = LocalCollider::new(
+            ObjectCollider::from(TriMesh::from(Cuboid::new(Vector::new(3., 3., 4.)))),
+            Isometry::new(point_msl.into(), Vector::identity()),
+        );
+
+        let entity = world
+            .spawn((
+                Transform::from_translation(point_msl),
+                Centre {
+                    x: point_msl.x,
+                    y: point_msl.y,
+                },
+                EnergyReceiver,
+                NearbyUnits::default(),
+                UnitNumber(i as u32),
+            ))
+            .id();
+
+        index.insert(entity, collider);
+    }
+
+    world.insert_resource(Clock(0.));
+    world.insert_resource(index);
+}
+
+/// Move entities in circles of radius DISTANCE_FROM_MAP_EDGE / 2.
+fn move_entities_in_circle(
+    clock: Res<Clock>,
+    mut query: Query<(&mut Transform, &UnitNumber, &Centre)>,
+) {
+    for (mut transform, unit_number, centre) in query.iter_mut() {
+        // Change direction (counter)clockwise based on entity_mum % 2 == 0
+        let direction = if unit_number.0 as f32 % 2. == 0. {
+            1.
+        } else {
+            -1.
+        };
+
+        let t = clock.0;
+        let radius = DISTANCE_FROM_MAP_EDGE / 2.;
+        let omega = SPEED / radius;
+        let omega_shift = unit_number.0 as f32;
+
+        let x = radius * (t * omega + omega_shift * direction).sin();
+        let y = radius * (t * omega + omega_shift * direction).cos();
+
+        transform.translation.x = x + centre.x;
+        transform.translation.y = y + centre.y;
+    }
+}
+
+fn nearby_benchmark(c: &mut Criterion) {
+    let mut group = c.benchmark_group("Nearby unit movement scenarios");
+
+    for i in [100, 1_000, 10_000, 100_000].iter() {
+        let mut app = App::default();
+        init_world_with_entities_moving(&mut app.world, *i);
+        app.add_systems(Update, update_nearby_recv);
+        app.add_plugins(TimePlugin);
+
+        let update_units_schedule = Schedule::default();
+        app.add_schedule(UpdateUnits, update_units_schedule);
+
+        app.add_systems(UpdateUnits, (update_index, move_entities_in_circle));
```

Comment:
I would run this two systems in chain so that the transform update **always** happens before the index update.